### PR TITLE
feat(childprocess/exec)

### PR DESCRIPTION
**Pull request by kaustavha**
First pass at integrating something like nodes exec which returns a callback with a buffer of all out&err data

### CODE REVIEW

File: deps/childprocess.lua [hunk 1]
```diff
@@ -164,4 +164,61 @@ local function spawn(command, args, options)
...
+  end
+  local function onClose(_exitCode)
+    exitCode = _exitCode
+    done()
+  end
+
+  local child = spawn(file, args, options)
+
+  child.stdout:on('data', function(chunk)
+    if #stdout > options.maxBuffer then
```
Comment (rphillips) [inline deps/childprocess.lua]: you will need to keep track of the number of bytes read for each chunk type. #stdout isn't going to give you the right number.

**Comment by rphillips** [comment]
Overall, we will need to use spawn() within the agent, since the overall output of a hostinfo might be over 200KB.

**Comment by kaustavha** [comment]
yes ryan im just doing this for the lulz, i think the hostinfos look good with the steams and the run function atm.

**Comment by rphillips** [comment]
The timeout support would be nice to have.

File: deps/childprocess.lua [hunk 2]
```diff
@@ -164,4 +164,71 @@ local function spawn(command, args, options)
...
+  end
+
+  if options.shell then file = options.shell end
+  if not options.maxBuffer then options.maxBuffer = 200 * 1024 end
+
+  -- http://wowwiki.wikia.com/wiki/Lua_object_memory_sizes
+  -- We're being generous here since the internal representation is to chunk them into a table
+  -- which have a base size of 40bytes but the final callback recieves a single string of each type
+  local stdoutchunkbytes = 24
+  local stderrchunkbytes = 24
```
Comment (creationix) [inline deps/childprocess.lua]: Why are we trying to take table size into account?  I would be very confused if I had set a max buffer size of 100 bytes and only got back 76 bytes.

File: deps/childprocess.lua [hunk 3]
```diff
@@ -164,4 +164,71 @@ local function spawn(command, args, options)
   return em
 end
 
+local function exec(command, args, options, callback)
+  local file
+  local stringArgs = command..table.concat(args, ' ')
```
Comment (creationix) [inline deps/childprocess.lua]: I think you're confusing `exec` with `execFile`.  In node, `exec` takes `(command, options, callback)` and runs the command in a subshell, but `execFile` takes `(file, args, options, callback)` and does not use a subshell.

File: deps/childprocess.lua [hunk 4]
```diff
@@ -164,4 +164,71 @@ local function spawn(command, args, options)
...
+  if process.env._system_name:lower() == 'win32' then
```
Comment (rphillips) [inline deps/childprocess.lua]: use los.type() here

File: deps/childprocess.lua (hunk 2 above)
Comment (rphillips) [inline deps/childprocess.lua]: also, we should chunk probably 4k at a minimum

File: deps/childprocess.lua (hunk 4 above)
Comment (creationix) [inline deps/childprocess.lua]: Also new luvi has `require('luvipath').isWindows` as a boolean.  This uses `jit.os == 'Windows'` on luajit and uses `not not package.path:match("\\")` as a fallback on other luas.

File: deps/childprocess.lua (hunk 3 above)
Comment (kaustavha) [inline deps/childprocess.lua]: You're right, but:
https://github.com/nodejs/node/blob/master/lib/child_process.js#L101-L107
So when I cycle back to this Ill port both

Comment (creationix) [inline deps/childprocess.lua]: Yep, `execFile` is lower level and is used by `exec`.  The important thing I wanted to note is that `exec` doesn't take args, but does inject `sh` or `cmd` while `execFile` is just spawn with output buffered.

**Comment by creationix** [comment]
You should add coroutine support to this to make it even easier to use from the repl.

See an example using the adapt helper at https://github.com/luvit/luvit/blob/master/deps/dns.lua#L696

**Comment by creationix** [comment]
To use the `adapt` helper you'll need to split `execFile` into two functions.  One to be called directly by the end-user and an internal one that does the actual work, but requires all arguments (no optional slots).

The two user-facing functions will normalize inputs and then call the real function via `adapt`.  You'll need to check for both `"function"` and `"thread"` when looking for the `callback` after optional callbacks since either functions or coroutine threads are possible with adapt.

File: deps/childprocess.lua [hunk 5]
```diff
@@ -164,4 +165,144 @@ local function spawn(command, args, options)
...
+    file = 'cmd.exe'
+    args = {'/s', '/c', '"'..stringArgs..'"' }
+  else
+    file = '/bin/sh'
+    args = {'-c', stringArgs }
+  end
+
+  if options and options.shell then file = options.shell end
+
+  return {
```
Comment (creationix) [inline deps/childprocess.lua]: Just a style nit, but I prefer multiple return values here over creating a temporary table that will simple be destructured on the other side.  Also it seems you're never using the `cmd` property returned here and it's just one of the input parameters anyway, the caller already knows it.

File: deps/childprocess.lua [hunk 6]
```diff
@@ -164,4 +165,144 @@ local function spawn(command, args, options)
...
+      table.insert(stderr, chunk)
+    end
+  end)
+
+  child:once('close', onClose)
+end
+
+local function exec(command, options, callback)
+  -- options is optional
+  local opts
```
Comment (creationix) [inline deps/childprocess.lua]: If you modify `normalizeExecArgs` to return multiple values, this entire function can be simplified to:

```lua
return execFile(normalizeExecArgs(command, callback and options, callback or options)
```

`callback and options` will always be `options` if callback is true or falsy if callback is falsy.

Comment (creationix) [inline deps/childprocess.lua]: Note that my proposed change preserves your logic.  Both will do the wrong thing if someone calls with an options table, but omits the callback.

File: deps/childprocess.lua [hunk 7]
```diff
@@ -164,4 +166,142 @@ local function spawn(command, args, options)
...
+    callback = function() end
+    options = {}
+    args = {}
+  elseif not options then
+    options = {}
+  elseif not args then
+    args = {}
+  end
+
+  adapt(callback, _exec, file, args, options)
```
Comment (creationix) [inline deps/childprocess.lua]: Need to tail call adapt for it to work (Return it's return value).

Comment (creationix) [inline deps/childprocess.lua]: With a return added here, I can verify it works.

```lua
> require('childprocess').exec("ls", thread)
'ChangeLog\nLICENSE.txt\nMakefile\nNOTICE.txt\nREADME.markdown\nVagrantfile\nappveyor.yml\nbench\ndeps\nexamples\ninit.lua\nmain.lua\nmake.bat\npackage.lua\nprompt.lua\ntests\ntools\n'	''
> 
```

**Comment by creationix** [comment]
+1

**Pull request merged**